xtask/src/nomt.rs

```diff
@@ -24,31 +27,40 @@ impl NomtDB {
 }
 
 impl DB for NomtDB {
-    fn apply_actions(&mut self, actions: Vec<Action>) {
+    fn apply_actions(&mut self, actions: Vec<Action>, timer: Option<&mut Timer>) {
+        let _timer_guard = timer.and_then(|t| Some(t.record()));
+
         let mut session = self.nomt.begin_session();
         let mut access: FxHashMap<KeyPath, KeyReadWrite> = FxHashMap::default();
 
         for action in actions.into_iter() {
             match action {
-                Action::Writes(writes) => {
-                    for (key, val) in writes {
-                        // TODO: Should the hash of the key be included in the benchmarking?
-                        let key_path = sha2::Sha256::digest(key).into();
-                        let value = val.map(|v| std::rc::Rc::new(v));
-
-                        match access.entry(key_path) {
-                            Entry::Occupied(mut o) => {
-                                o.get_mut().write(value);
-                            }
-                            Entry::Vacant(v) => {
-                                v.insert(KeyReadWrite::Write(value));
-                            }
-                        }
+                Action::Write { key, value } => {
+                    let key_path = sha2::Sha256::digest(key).into();
+                    let value = value.map(|v| std::rc::Rc::new(v));
 
-                        session.tentative_write_slot(key_path);
+                    match access.entry(key_path) {
+                        Entry::Occupied(mut o) => {
+                            o.get_mut().write(value);
+                        }
+                        Entry::Vacant(v) => {
+                            v.insert(KeyReadWrite::Write(value));
+                        }
                     }
+
+                    session.tentative_write_slot(key_path);
+                }
+                Action::Read { key } => {
+                    let key_path = sha2::Sha256::digest(key).into();
+                    let _value = match access.entry(key_path) {
+                        Entry::Occupied(o) => o.get().last_value().cloned(),
+                        Entry::Vacant(v) => {
+                            let value = session.tentative_read_slot(key_path).unwrap();
+                            v.insert(KeyReadWrite::Read(value.clone()));
+                            value
+                        }
+                    };
                 }
```

Comment:
Where does the commit take place?

Reply:
At the end of the loop over all actions:

```rust
        let mut actual_access: Vec<_> = access.into_iter().collect();
        actual_access.sort_by_key(|(k, _)| *k);
        self.nomt.commit_and_prove(session, actual_access).unwrap();
```